qt/python/mantidqt/mantidqt/plotting/sample_shape.py

```diff
@@ -0,0 +1,424 @@
+    # Set limits to new limits, and replace with old limits that are wider
+    if old_limits:
+        min_x = compare_and_replace_minimum(new_limits[0], old_limits[0])
+        min_y = compare_and_replace_minimum(new_limits[1], old_limits[1])
+        min_z = compare_and_replace_minimum(new_limits[2], old_limits[2])
+        max_x = compare_and_replace_maximum(new_limits[3], old_limits[3])
+        max_y = compare_and_replace_maximum(new_limits[4], old_limits[4])
+        max_z = compare_and_replace_maximum(new_limits[5], old_limits[5])
+    else:
+        [min_x, min_y, min_z, max_x, max_y, max_z] = new_limits
+    return [min_x, min_y, min_z, max_x, max_y, max_z]
+
+
+def calculate_beam_direction(source, sample):
+    source_position = source.getPos()
+    sample_position = sample.getPos()
+    beam_vector = sample_position - source_position
+    return beam_vector
+
+
+def add_beam_arrow(plot_axes, workspace):
+    # Add arrow along beam direction
+    source = workspace.getInstrument().getSource()
+    sample = workspace.getInstrument().getSample()
+    if source and sample:
+        beam_origin = plot_axes.get_xlim3d()[0], plot_axes.get_ylim3d()[0], plot_axes.get_zlim3d()[0]
+        beam_direction = calculate_beam_direction(source, sample)
+        add_arrow(plot_axes, beam_direction, origin=beam_origin)
+
+
+def add_arrow(ax, vector, origin=None, factor=None, color='black', linestyle='-'):
+    # Add arrows for Beam or Crystal lattice
+    if origin is None:
+        origin = (ax.get_xlim3d()[1], ax.get_ylim3d()[1], ax.get_zlim3d()[1])
+    if factor is None:
+        lims = ax.get_xlim3d()
+        factor = (lims[1]-lims[0]) / 3.0
+    vector_norm = vector / np.linalg.norm(vector)
+    ax.quiver(
+        origin[0], origin[1], origin[2],
+        vector_norm[0]*factor, vector_norm[1]*factor, vector_norm[2]*factor,
+        color=color,
+        linestyle=linestyle
+    )
+
+
+def plot_lattice_vectors(plot_axes, workspace):
+    """Add arrows for real and reciprocal lattice vectors"""
+    real_lattice_vectors, reciprocal_lattice_vectors = calculate_lattice_vectors(workspace)
+    colors = ['r', 'g', 'b']
+    plot_real_lattice_vectors(plot_axes, real_lattice_vectors, colors)
+    plot_reciprocal_lattice_vectors(plot_axes, reciprocal_lattice_vectors, colors)
+
+
+def plot_reciprocal_lattice_vectors(plot_axes, reciprocal_lattice, colors):
+    for i in range(3):  # plot reciprocal_lattice with '--' dashed linestyle
+        add_arrow(plot_axes, reciprocal_lattice[:, i], color=colors[i], linestyle='--')
+
+
+def plot_real_lattice_vectors(plot_axes, real_lattice, colors):
+    for i in range(3):  # plot real_lattice with '-' solid linestyle
+        add_arrow(plot_axes, real_lattice[:, i], color=colors[i])
+
+
+def calculate_lattice_vectors(workspace):
+    ub_matrix = np.array(workspace.sample().getOrientedLattice().getUB())
+    hkl = np.array([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]])
```

Comment:
Perhaps worth a comment here that the lattice vectors are in columns. Not sure the numpy syntax helps where because it suggests they're in rows. Later on you pull out individual columns when plotting the arrows

Reply:
This code is from the original scripts, so I was looking back at my Slack convo with Richard. And he had mentioned that the lattice vectors such as b1 are column vectors. But I guess I was meaning hkl to be rows. Is there a better way I can represent the code?